File: head_to_offer/28/hto_28.py
from typing import List, Dict
from queue import Queue
from collections import deque
import math


class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None
    
class Solution:
    def isSymmetric(self, root: TreeNode) -> bool:
        if not root: 
            return True
        queue = deque()
        queue.append(root.left)
        queue.append(root.right)
        while queue:
            first = queue.popleft()
            second = queue.popleft()
            if not first and not second: 
                continue
            if not first or not second or first.val != second.val: 
                return False
            queue.append(first.left)
            queue.append(second.right)
            queue.append(first.right)
            queue.append(second.left)
        return True


if __name__ == "__main__":
    node1 = TreeNode(1)
    node2 = TreeNode(2)
    node3 = TreeNode(2)
    node4 = TreeNode(3)
    node5 = TreeNode(3)
    node1.left = node2
    node1.right = node3
    node2.left = node4
    node3.right = node5
    s = Solution()
    print(s.isSymmetric(node1))

File: head_to_offer/28/hto_28.cpp
#include <iostream>
#include <string>
#include <vector>
#include <climits>
#include <algorithm>
#include <cstring>
#include <unordered_map>
#include <unordered_set>
#include <map>
#include <stack>
#include <bitset>
#include <array>
#include <queue>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x): val(x), left(NULL), right(NULL) { }
};

class Solution {
public:
    bool isSymmetric(TreeNode *root) {
        if (!root) return true;
        TreeNode *first;
        TreeNode *second;
        queue<TreeNode *> qq;
        qq.push(root->left);
        qq.push(root->right);
        while (!qq.empty()) {
            first = qq.front();
            qq.pop();
            second = qq.front();
            qq.pop();
            if (!first && !second) continue;
            if (!first || !second || first->val != second->val) return false;
            qq.push(first->left);
            qq.push(second->right);
            qq.push(first->right);
            qq.push(second->left);
        }
        return true;
    }
};

int main() {
    TreeNode *node1 = new TreeNode(1);
    TreeNode *node2 = new TreeNode(2);
    TreeNode *node3 = new TreeNode(2);
    TreeNode *node4 = new TreeNode(3);
    TreeNode *node5 = new TreeNode(3);
    node1->left = node2; node1->right = node3;
    node2->left = node4; node3->right = node5;
    Solution s;
    cout << boolalpha << s.isSymmetric(node1) << endl;
}

